src/Evolution/DgSubcell/SliceData.cpp

```diff
@@ -0,0 +1,151 @@
+  for (size_t j = lower_bounds[1], index = 0; j < upper_bounds[1]; ++j) {
+    for (size_t i = lower_bounds[0]; i < upper_bounds[0]; ++i, ++index) {
+      (*sliced_subcell_vars)[component_offset + index] =
+          volume_subcell_vars[component_offset_volume + i +
+                              j * volume_extents[0]];
+    }
+  }
+}
+
+template <>
+void copy_data(const gsl::not_null<std::vector<double>*> sliced_subcell_vars,
+               const gsl::span<const double>& volume_subcell_vars,
+               const size_t component_offset,
+               const size_t component_offset_volume,
+               const std::array<size_t, 3>& lower_bounds,
+               const std::array<size_t, 3>& upper_bounds,
+               const Index<3>& volume_extents) noexcept {
+  for (size_t k = lower_bounds[2], index = 0; k < upper_bounds[2]; ++k) {
+    for (size_t j = lower_bounds[1]; j < upper_bounds[1]; ++j) {
+      for (size_t i = lower_bounds[0]; i < upper_bounds[0]; ++i, ++index) {
+        (*sliced_subcell_vars)[component_offset + index] =
+            volume_subcell_vars[component_offset_volume + i +
+                                volume_extents[0] *
+                                    (j + k * volume_extents[1])];
+      }
+    }
+  }
+}
+}  // namespace
+
+template <size_t Dim>
+DirectionMap<Dim, std::vector<double>> slice_data_impl(
+    const gsl::span<const double>& volume_subcell_vars,
+    const Index<Dim>& subcell_extents, const size_t number_of_ghost_points,
+    const DirectionMap<Dim, bool>& directions_to_slice) noexcept {
+  const size_t num_pts = subcell_extents.product();
+  const size_t number_of_components = volume_subcell_vars.size() / num_pts;
+  std::array<size_t, Dim> result_grid_points{};
+  for (size_t d = 0; d < Dim; ++d) {
+    gsl::at(result_grid_points, d) =
+        number_of_ghost_points * subcell_extents.slice_away(d).product();
+  }
+  DirectionMap<Dim, std::vector<double>> result{};
+  for (size_t d = 0; d < Dim; ++d) {
+    if (directions_to_slice.count(Direction<Dim>(d, Side::Lower)) == 1 and
+        directions_to_slice.at(Direction<Dim>(d, Side::Lower))) {
+      result[Direction<Dim>{d, Side::Lower}] = std::vector<double>(
+          gsl::at(result_grid_points, d) * number_of_components);
+    }
+    if (directions_to_slice.count(Direction<Dim>(d, Side::Upper)) == 1 and
+        directions_to_slice.at(Direction<Dim>(d, Side::Upper))) {
+      result[Direction<Dim>{d, Side::Upper}] = std::vector<double>(
+          gsl::at(result_grid_points, d) * number_of_components);
+    }
+  }
+
+  for (size_t component_index = 0; component_index < number_of_components;
+       ++component_index) {
+    const size_t component_offset_volume = component_index * num_pts;
+    for (size_t d = 0; d < Dim; ++d) {
+      const size_t component_offset_result =
+          gsl::at(result_grid_points, d) * component_index;
+      if (result.count(Direction<Dim>(d, Side::Lower)) == 1) {
```

Comment:
Same as above,
- could this be a loop over `result` instead of a loop over `d`?
- if it has to be a loop over `d`, why not use `.contains()`?

Reply:
The thought process was cache-locality, but I haven't measured anything (and I'd be surprised if this is very expensive either way). I'll switch to the simpler loops and we can always re-evaluate if we find a perf issue :)